### PR TITLE
Add listEnvironments DDB operation

### PR DESCRIPTION
**Pull request by gongmax**
#### Summary
This is the DDB level implementation of listEnvironments. Cluster.java in dataServiceModel is stolen from blox/blox#323.

#### Testing
gradlew build
gradlew check

New tests cover the changes: yes

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+@Data
+@Builder
+// required for builder
+@AllArgsConstructor
+// required for mapstruct
+@NoArgsConstructor
+public class Cluster {
+  @NonNull private String accountId;
+  @NonNull private String cluster;
+}
```
Comment (wbingli) [inline data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java]: I'm thinking whether github PR could have code review diff not have to against with the target branch. Like this case,  you may want to only have diff for your change without the Cluster change.  Otherwise, you will have to resolve conflict later on.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.NonNull;
+
+@Data
+@Builder
+// required for builder
+@AllArgsConstructor
+// required for mapstruct
+@NoArgsConstructor
+public class Cluster {
+  private static final String UNDERSCORE = "_";
```
Comment (wbingli) [inline data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: Not in scope of this PR, but not sure using underscore as delimiter is good idea, that means we will not allow cluster and environment name to have underscore.   

It will hurt customer as lots of people like to use underscore to name stuff.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java [hunk 3]
```diff
@@ -32,6 +33,8 @@ Environment createEnvironmentAndEnvironmentRevision(
   Environment getEnvironment(EnvironmentId environmentId)
       throws ResourceNotFoundException, InternalServiceException;
 
+  List<Environment> listEnvironments(Cluster cluster) throws InternalServiceException;
```
Comment (wbingli) [inline data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: Should also throw `ResourceNotFoundException` for cluster not found?

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDB.java [hunk 4]
```diff
@@ -306,6 +307,25 @@ public Environment getEnvironment(@NonNull final EnvironmentId environmentId)
...
+      throws InternalServiceException {
+    try {
+      return dynamoDBMapper
+          .query(
+              EnvironmentDDBRecord.class,
+              new DynamoDBQueryExpression<EnvironmentDDBRecord>()
+                  .withHashKeyValues(
+                      EnvironmentDDBRecord.withHashKeys(cluster.generateAccountIdCluster())))
+          .stream()
+          .map(e -> environmentMapper.toEnvironment(e))
```
Comment (wbingli) [inline data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDB.java]: Should be able to use method reference `environmentMapper::toEnvironment`

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java [hunk 5]
```diff
@@ -42,6 +42,10 @@
   public static final String ENVIRONMENT_NAME_RANGE_KEY = "environmentName";
   public static final String LATEST_ENVIRONMENT_REVISION_ID = "latestEnvironmentRevisionId";
 
+  public static EnvironmentDDBRecord withHashKeys(final String accountIdCluster) {
+    return EnvironmentDDBRecord.builder().accountIdCluster(accountIdCluster).build();
```
Comment (wbingli) [inline data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java:46]: units?

File: data-service/src/test/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDBTest.java [hunk 6]
```diff
@@ -124,4 +141,44 @@ public void testGetEnvironmentRevisionInternalError() throws Exception {
...
+  @Test
+  public void testListEnvironments() throws Exception {
+    when(environmentDDBRecords.stream()).thenReturn(Stream.of(environmentDDBRecord));
+    when(dynamoDBMapper.query(eq(EnvironmentDDBRecord.class), any(DynamoDBQueryExpression.class)))
+        .thenReturn(environmentDDBRecords);
+    when(environmentMapper.toEnvironment(environmentDDBRecord)).thenReturn(environment);
+
+    final List<Environment> result = environmentRepositoryDDB.listEnvironments(cluster);
+
+    verify(dynamoDBMapper).query(eq(EnvironmentDDBRecord.class), any());
```
Comment (wbingli) [inline data-service/src/test/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDBTest.java]: It would be better to capture the DynamoDBQueryExpression and you can verify the passed hash key:
```
verify(dynamoDBMapper).query(eq(EnvironmentDDBRecord.class), queryExpCaptor.capture());
assertThat(queryExpCaptor.getValue().getHashKeyValues().getAccountIdCluster()).isEqualsTo(CLUSTER. generateAccountIdCluster())
```

File: frontend-service/api/swagger.yml [hunk 7]
```diff
@@ -428,4 +428,4 @@ definitions:
         type: "string"
       newTargetRevisionId:
         type: "string"
-x-generated-at: "2018-01-09T22:55:42.535Z"
+x-generated-at: "2018-01-09T22:55:42.535Z"
```
Comment (xjhe) [inline frontend-service/api/swagger.yml]: Do we need to update this line?

Comment (wbingli) [inline frontend-service/api/swagger.yml]: why this change?  I would think having a new line at the end of file is good practice.  You can enable the "Ensure line feed at file end on save" if using intellij.

File: data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java (hunk 1 above)
Comment (gongmax) [reply data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java]: I copied PR 323 only because this can unblock some change in my PR. And since I copied it, I don't think there will be conflict.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java (hunk 2 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: Good catch, but I don't think this disallow customer to use underscore in their cluster/environment name. We only use the underscore to form the hashkey in DDB, and it seems we do not need to get the cluster/environment name/id from the hashkey.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java (hunk 3 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: What do you mean by "cluster not found"? If there is no record associated with this cluster and account in DDB, this api just return an empty list.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDB.java (hunk 4 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDB.java]: yeah will update

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java (hunk 5 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java:46]: Can you explain more?

File: data-service/src/test/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDBTest.java (hunk 6 above)
Comment (gongmax) [reply data-service/src/test/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepositoryDDBTest.java]: will update

File: frontend-service/api/swagger.yml (hunk 7 above)
Comment (gongmax) [reply frontend-service/api/swagger.yml]: I don't know. This is auto-generated, and I don't know what change here is. It seems to me there is no code change here and no line added or deleted.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java (hunk 2 above)
Comment (xjhe) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: I think the actual place that this underscore will matter is in the EnvironmentId object? Because we do try to decompose the accountID and cluster from the key. We may think about the impact there. https://github.com/blox/blox/blob/dev/data-service/src/main/java/com/amazonaws/blox/dataservice/model/EnvironmentId.java#L46

Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: Oh you are right. We should reconsider this.

File: frontend-service/api/swagger.yml (hunk 7 above)
Comment (wjbuys) [reply frontend-service/api/swagger.yml]: This is auto-updated by [this filter](https://github.com/blox/blox/blob/dev/buildSrc/src/main/java/com/amazonaws/blox/swagger/GenerationTimestampFilter.java), which seemed like a good idea at the time, but now seems like it just makes annoying noise in our diffs.

I think we can remove that filter in a later PR to avoid this issue in future.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java (hunk 2 above)
Comment (wjbuys) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: Agreed. I'd say we should use the same separators as ARNs, so identifiers that are valid in ARNs will also be valid in our DB key.

File: data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java (hunk 1 above)
Comment (wjbuys) [reply data-service-model/src/main/java/com/amazonaws/blox/dataservicemodel/v1/model/Cluster.java]: Sorry about that, I forgot to actually push that yesterday. I think you'll be able to rebase your PR on that change now.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java (hunk 3 above)
Comment (wbingli) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: I think they are two different things. 

- Like ResouceNotFound for cluster, which means no such clsuter defined. 
- Emtpy list means there is no environment in this cluster, but cluster does exist.

My idea is following the same rule of list api, say list-services:
```
 aws ecs list-services --cluster not-exist

An error occurred (ClusterNotFoundException) when calling the ListServices operation: Cluster not found.
```

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java (hunk 5 above)
Comment (wbingli) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java:46]: oh, just mention we should add unit test for the code change.   It's obvious but it won't hurt to have a simple unit test.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java (hunk 3 above)
Comment (wbingli) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: The implementation would be different as we don't track the cluster.  Would be a front-end service validation thing to validate the cluster with ECS before invoke data service.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java (hunk 2 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/model/Cluster.java:32]: So we will use cluster name not cluster arn across all of our apis and ddb tables?

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java (hunk 3 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: I agree we need validation for non-exist cluster but I don't think we should take care of it at this level. I'm leaning to add it in data-service api rather than this DDB interacting level.

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java (hunk 5 above)
Comment (gongmax) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/model/EnvironmentDDBRecord.java:46]: Fixed

File: data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java (hunk 3 above)
Comment (wbingli) [reply data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java:36]: yes, the cluster validation should be on higher level

**Review by wbingli — APPROVED** [review]
LGTM
